repo2docker/buildpacks/conda/__init__.py

```diff
@@ -178,7 +178,19 @@ def py2(self):
         """Am I building a Python 2 kernel environment?"""
         return self.python_version and self.python_version.split(".")[0] == "2"
 
-    def get_assemble_scripts(self):
+    def get_preassemble_script_files(self):
+        """preassembly only requires environment.yml
+
+        enables caching assembly result even when
+        repo contents change
+        """
+        assemble_files = super().get_preassemble_script_files()
+        environment_yml = self.binder_path("environment.yml")
+        if os.path.exists(environment_yml):
+            assemble_files[environment_yml] = environment_yml
```

Comment:
Should we attempt to sort the contents of a `environment.yml` so that two repos asking for the same packages in a different order get a cache hit? Or is this a level two optimisation for a new PR?

Reply:
ooh, that's a very interesting question. I think that's a job for a second PR. I'm also not sure how often this is likely to happen, since it has to be a truly identical environment specification (not just equivalent) unless we import conda and start resolving dependencies, which would slow down builds quite a bit, I think, since it would have to happen prior to generating the Dockerfile.

Before spending much time on that, I think it'd be worth trying to survey environment.ymls in the wild and seeing how many:

1. are not identical byte-wise, but
2. are identical, dependency-wise

Reply:
Created an issue to tackle this https://github.com/jupyter/repo2docker/issues/749